### ISSUE TITLE
Support request for basic.recover

### ISSUE THREAD

**Issue by pkruithof** [pdezwart/php-amqp#138]
It seems that the [basic.recover][1] command is not available in this extension. Could this be implemented?

As a related question: would it be a good idea to add some generic method to send a command to the AMQP server? As a way to be able to use commands that have not yet been implemented. 

[1]: http://www.rabbitmq.com/amqp-0-9-1-quickref.html#basic.recover

**Comment by pinepain** [comment]
Yeah, `basic.recover` is not implemented yet. php-amqp is not 100% AMQP protocol implementation yet and there are other methods which are not yet supported.

I think [basic.recover](http://www.rabbitmq.com/amqp-0-9-1-quickref.html#basic.recover) may be implemented without any side-effects to existent code.

`basic.recover` method can be placed to AMQPChannel method (while it breaks single responsibility principle, it is the most logical place for channel-level method). Also, errata notes about [basic.recover with requeue=false](http://www.rabbitmq.com/amqp-0-9-1-errata.html#section_2) and [basic.recover synchronicity](http://www.rabbitmq.com/amqp-0-9-1-errata.html#section_28) should be kept in mind. In addition to errata, [according to RabbitMQ specs](http://www.rabbitmq.com/specification.html) `basic.recover` has partial support (Recovery with requeue=false is not supported.)

I will take a look on this issue.

**Comment by pkruithof** [comment]
Thank you, much appreciated.

Could you give your thoughts on the other question?

> would it be a good idea to add some generic method to send a command to the AMQP server? As a way to be able to use commands that have not yet been implemented.

**Comment by pinepain** [comment]
Technically it is possible but I don't think that it is good idea. At least for now. 

Can you explain purposes to allow custom AMQP methods sending and handling in userspace (except adding missed AMQP methods support)?

**Comment by pinepain** [comment]
Technically it is possible but I don't think that it is good idea. At least for now.

Can you explain purposes to allow custom AMQP methods sending and handling in userspace (except adding missed AMQP methods support)?

TL;DR

    1. Different methods has their own signature as well as different kind of responses. Generally, yes, it is possible to have some generic method to send a command to AMQP server. The bottom line is that users have to handle response by their own and that they have to know AMQP protocol pretty well while some methods requires response, some of them not, some of them sends data in response, some of them not always.
    2. Current php-amqp architecture is a bit messy: AMQP protocol methods organized in a bit inconsistent and sometimes AMQP methods support is not full or partly incorrect.
    3. There are some general conflicts between how it is designed (in synchronous way, like PHP itself) and how AMQP protocol designed (mostly, asynchronous) which leads to some potential problems (like catching in response to one method exception that was thrown for another. While it is a bit rare case, the whole assumption that we always get in response data related to latest request is generally invalid (vast majority of AMQP methods IS asynchronous).

**Comment by pinepain** [comment]
Technically it is possible but I don't think that it is good idea. At least for now.

Can you explain purposes to allow custom AMQP methods sending and handling in userspace (except adding missed AMQP methods support)?

TL;DR

 1. Different methods has their own signature as well as different kind of responses. Generally, yes, it is possible to have some generic method to send a command to AMQP server. The bottom line is that users have to handle response by their own and that they have to know AMQP protocol pretty well while some methods requires response, some of them not, some of them sends data in response, some of them not always.
 2. Current php-amqp architecture is a bit messy: AMQP protocol methods organized in a bit inconsistent and sometimes AMQP methods support is not full or partly incorrect.
 3. There are some general conflicts between how it is designed (in synchronous way, like PHP itself) and how AMQP protocol designed (mostly, asynchronous) which leads to some potential problems (like catching in response to one method exception that was thrown for another. While it is a bit rare case, the whole assumption that we always get in response data related to latest request is generally invalid (vast majority of AMQP methods IS asynchronous).

**Comment by pkruithof** [comment]
The purpose was mainly to have a way to manually perform a method that is not yet implemented. Like right now, I kind of need the `basic.recover` method, but I'm unable to do so (unless I use another library).

But your arguments make sense to not do it. It also makes me wonder about the future development about this extension. Do you think it's subject to change a lot due to the 'messy architecture' and design approach (asynch. vs synch.)? In other words, should I be concerned that the public API of this extension will have a major BC break sometime in the future?

**Comment by pinepain** [comment]
I will add `basic.recover` ASAP and will review all supported methods to be sure we didn't missed anything.

Right now I'm fixing memleaks and minor errors introduced after huge refactoring (we didn't released that version to public). ETA is about 24 hours while most changes is production-critical.

As to BC, thinks should be discussed with other contributors and users, but in short terms I would propose further steps:
 0. Provide high-level abstraction which provide unified interface to both [php-amqplib library](https://github.com/videlalvaro/php-amqplib) and [php-amqp extension](https://github.com/pdezwart/php-amqp) and probably for any other implementations (if any or if there will any kind of them). I've already done some researches and PoC on this, but didn't have much free time, so ETA for development version may vary from few days to few month or even years. There was talks about unifying both php-amqplib and php-amqp interface, but it is not so easy as there are already large established userbase for both projects.
 1. Keep BC inside current  version (1.*) as long as possible and as long as users use it for at least 1 year (until, say Q1 2016) and after that backport security or performance fixes for about 2 years (say, to Q1 2018). These terms are more than enough even for moving legacy code base to newer version or mimic  interfaces to php-amqp with high-level abstraction (just define classes with same interfaces like now and proxy method calls to new implementation). This would be fair deal for those who can't upgrade their codebase (due to numerous reasons).
 2. Roll out redesigned php-amqp extension, let it call php-amqp2 for now, and then add it support to abstraction, so end-users will not be affected by changes.

Also I have in plans to add AMQP support for most popular frameworks (Symfony, Zend, Laravel, etc) to offer their users intuitive interface (in terms of particular framework) to deal with AMQP-related stuff.
 
An important part of development process it also required to provide manuals and examples (which we have lack of for now, actually, the only manual available is stubs shipped with extension).

As you see, there are a lot of things to do when the community is pretty small, so any help, even discussion will be appreciated.

// cc @lstrojny, @bkw

**Comment by pkruithof** [comment]
Wow, much more than I hoped for, thank you!

Regarding the unification of interfaces, maybe it's wise to look at the [Java implementation][1]. It's been around for longer, I think, and probably has figured out solutions for the same kind of problems.

My desired approach would be to indeed unify the interfaces, and create a version 2 out of that. That way there will be 1 well designed implementation which we can move forward.

[1]: http://www.rabbitmq.com/javadoc/overview-summary.html

I'm not a C programmer so there's little I can do to help in terms of coding. But I'll try to keep an eye on other aspects where I can chip in (documentation maybe?).

**Comment by lstrojny** [comment]
@pinepain I agree on the BC part very much, the only larger API changes we did so far in the 1.x cycle was splitting timeout into read- and write-timeout and declare vs. declareExchange and declareQueue. That was OK and trivial to be compatible with (just renamed or splits).

On a potential v2: any particular API changes you would have in mind already?

**Comment by pinepain** [comment]
@pkruithof `basic.recover` support added, see https://github.com/pdezwart/php-amqp/pull/145. You can build against [basic_recover](https://github.com/pinepain/php-amqp/tree/basic_recover) branch.

To emit `basic.recover`, call `AMQPChannel::basicRecover()` method.

Note, that `basic.recover` **may redeliver zero or more messages**. For example, if you have only one consumer and you call `basic.recover` on the same channel with that consumer, zero messages will be redelivered.

As for me, `basic.recover` usage is quite limited to specific cases like RPC over AMQP and other tasks about highly reliable services. Also, if you will close channel, all unacked messages will be then scheduled back to the queue.

**Comment by pkruithof** [comment]
Thank you!

> For example, if you have only one consumer and you call basic.recover on the same channel with that consumer, zero messages will be redelivered.

Could you explain why this is?

**Comment by pinepain** [comment]
I was a bit wrong, sorry.

It *will looks like* no messages will be redeliverd in case of single consumer if you will still have that consumer stay active because messages will be redelivered back to that consumer. If you will have other consumers, even then some messages will be redelivered back to original consumer that didn't ack that message first.
